src/components/TableView/TableView.tsx

```diff
@@ -0,0 +1,182 @@
+import React, { FC, useState, memo } from "react";
+
+import Avatar from "@material-ui/core/Avatar";
+import Checkbox from "@material-ui/core/Checkbox";
+import Paper from "@material-ui/core/Paper";
+import Table, { TableProps } from "@material-ui/core/Table";
+import TableBody from "@material-ui/core/TableBody";
+import TableCell from "@material-ui/core/TableCell";
+import TableContainer from "@material-ui/core/TableContainer";
+import TableHead from "@material-ui/core/TableHead";
+import TablePagination from "@material-ui/core/TablePagination";
+import TableRow from "@material-ui/core/TableRow";
+
+import TableToolbar from "./TableToolbar";
+import { useStyles } from "./TableView.styles";
+import { ColumnsOptionProps } from "./TableView.types";
+import { Rows } from "./TableView.types.example";
+
+interface TableViewProps {
+  title?: string;
+  columns: (string | ColumnsOptionProps)[];
+  rows: Rows[];
+  tableProps?: TableProps;
+  // paginationProps starts here
+  withPagination?: boolean;
+  onPaginationChange?: (newPage: number) => void;
+  rowsPerPage?: number;
+  count?: number;
+  // paginationProps ends here
+  withSearch?: boolean;
+  withCheckbox?: boolean;
+}
+
+const TableView: FC<TableViewProps> = ({
+  title = "Table Example",
+  columns,
+  rows,
+  tableProps,
+  withPagination = false,
+  onPaginationChange = () => {
+    return null;
+  },
+  rowsPerPage = 0,
+  count = 0,
+  withSearch = false,
+  withCheckbox = false,
+}) => {
+  const [page, setPage] = useState(0);
+  const [selected, setSelected] = useState<string[]>([]);
+
+  const classes = useStyles();
+
+  const handleSelectAllClick = (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (event.target.checked) {
+      const newSelecteds = rows.map(n => n.email);
+      setSelected(newSelecteds);
+      return;
+    }
+    setSelected([]);
+  };
+
+  const handleClick = (event: React.MouseEvent<unknown>, name: string) => {
+    const selectedIndex = selected.indexOf(name);
+    let newSelected: string[] = [];
+
+    if (selectedIndex === -1) {
+      newSelected = newSelected.concat(selected, name);
+    } else if (selectedIndex === 0) {
+      newSelected = newSelected.concat(selected.slice(1));
+    } else if (selectedIndex === selected.length - 1) {
+      newSelected = newSelected.concat(selected.slice(0, -1));
+    } else if (selectedIndex > 0) {
+      newSelected = newSelected.concat(
+        selected.slice(0, selectedIndex),
+        selected.slice(selectedIndex + 1)
+      );
+    }
```

Comment:
How do you think if we can update above code like below?
```
const selectedIndex = selected.indexOf(name);
if (selectedIndex > -1) {
    selectedIndex.splice(selectedIndex, 1);
} else {
    newSelected = [... newSelected, name]
}
```

Reply:
`selectedIndex ` is a number so it won't be able to `splice` it.

if you meant like below, the behavior will change to be able to select only one row at a time.

```js
const selectedIndex = selected.indexOf(name);
if (selectedIndex > -1) {
    newSelected.splice(selectedIndex, 1);
} else {
    newSelected = [...newSelected, name]
}
```

Sorry if I misunderstood.

Reply:
@yukinoda `selectedIndex` value is returned by indexOf function so it's always a number. A apart from that, `handleClick` is called when you click on a checkbox of each row so we just handle only 1 row at a time.

Reply:
Haven't really understood it.

```diff
--- a/src/components/TableView/TableView.tsx
+++ b/src/components/TableView/TableView.tsx
@@ -64,17 +64,10 @@ const TableView: FC<TableViewProps> = ({
     const selectedIndex = selected.indexOf(key);
     let newSelected: string[] = [];
 
-    if (selectedIndex === -1) {
-      newSelected = newSelected.concat(selected, key);
-    } else if (selectedIndex === 0) {
-      newSelected = newSelected.concat(selected.slice(1));
-    } else if (selectedIndex === selected.length - 1) {
-      newSelected = newSelected.concat(selected.slice(0, -1));
-    } else if (selectedIndex > 0) {
-      newSelected = newSelected.concat(
-        selected.slice(0, selectedIndex),
-        selected.slice(selectedIndex + 1)
-      );
+    if (selectedIndex > -1) {
+      selected.splice(selectedIndex, 1);
+    } else {
+      newSelected = [...newSelected, key];
     }
 
     setSelected(newSelected);
```

 If doing the above changes, I will be able to only select one row at a time
https://user-images.githubusercontent.com/16083322/103846887-77f04c80-50e2-11eb-9042-6ab0ef12d322.mov

instead of multi-selection
![image](https://user-images.githubusercontent.com/16083322/103847041-bd147e80-50e2-11eb-9ecb-63207600e41e.png)

Reply:
@yukinoda correct me if I'm wrong. You're having 2 separate functions to handle row selecting: `handleSelectAllClick` to select all row and `handleClick` to toogle select/unselect each single row, right?

Reply:
@mlh-binhnt right

Reply:
@yukinoda Oops, sorry. I've updated the code. Please try again
if (selectedIndex > -1) {
      selected.splice(selectedIndex, 1);
      newSelected = [...selected];
    } else {
      newSelected = [...selected, key];
    }

Reply:
@mlh-binhnt Looks good and cleaner than the original one! Thanks! We'll make the commit.

Reply:
@mlh-binhnt 
Done 56b22b4
Please mark this as resolve if it looks good to you.